### PR TITLE
integration tests: python machine lib

### LINKED ISSUE #783: Add integration test for bluechictl metrics listen
Add integration test, which will verify `bluechictl metrics listen` functionality

### PR DESCRIPTION
**Pull request by mkemel**
Adding python machine lib with bluechictl background runner for integration tests. This allows sharing code between python scripts run on testing machine (container/ssh).

Also adding `bluechictl metrics listen` test

Fixes: #783

### CODE REVIEW

File: tests/bluechi_test/machine.py [hunk 1]
```diff
@@ -122,6 +122,21 @@ def copy_container_script(self, script_file_name: str):
              with content:\n{content}")
         self.create_file(target_dir, script_file_name, content)
 
+    def copy_machine_lib(self):
+        LOGGER.info(f"Copying machine lib...")
+        curr_dir = os.getcwd()
+        source_dir = os.path.join(curr_dir, "..", "..", "..", "bluechi_test", "bluechilib")
+        if not os.path.isdir(source_dir):
+            LOGGER.info("bluechilib directory not found, proceeding")
```
Comment (mwperina) [inline tests/bluechi_test/machine.py:132]: Shouldn't it be a warning?

File: tests/bluechi_test/bluechilib/bluechictl.py
Comment (mwperina) [inline tests/bluechi_test/bluechilib/bluechictl.py:1]: Isn't it confusing to have `bluechilib`, which should be used only on controller node, inside `bluechi_test`, which should be used only on test runner machine? Wouldn't it be better to have `bluechilib` in the same directory as `bluechi_test`? And maybe rename `bluechilib` to something like `bluechi_node_test`?

File: tests/bluechi_test/bluechilib/bluechictl.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        with tempfile.NamedTemporaryFile() as out_file:
+            try:
+                self.bluechictl_proc = subprocess.Popen(self.command, stdout=out_file, bufsize=1)
+
+                with FileFollower(out_file.name) as bluechictl_out:
+                    while self.bluechictl_proc.poll() is None and not self.finished:
+                        for line in bluechictl_out:
+                            print(f"Evaluating line '{line}'")
+                            match = None
+                            for pattern in copy.copy(self.expected_patterns):
```
Comment (mwperina) [inline tests/bluechi_test/bluechilib/bluechictl.py]: Just evaluating list of matches is not enough for more complicated use cases such as

https://github.com/eclipse-bluechi/bluechi/blob/main/tests/tests/tier0/bluechi-monitor-service/python/monitor.py
https://github.com/eclipse-bluechi/bluechi/blob/main/tests/tests/tier0/bluechictl-monitor-unit-created-removed/python/monitor.py

I suggest to use following class to evaluate incoming lines:

```python
class EventEvaluator:

def process_line(self, line: str) -> None:

def processing_finished(self) -> Bool:
```

Then each test could implement its own ancestor of this class.

File: tests/bluechi_test/bluechilib/bluechictl.py
Comment (mkemel) [reply tests/bluechi_test/bluechilib/bluechictl.py:1]: I originally named it `bluechi_machine_lib` - machine being container/vm/baremetal. WDYT?

File: tests/bluechi_test/bluechilib/bluechictl.py (hunk 2 above)
Comment (mkemel) [reply tests/bluechi_test/bluechilib/bluechictl.py]: I thought about something like this, just thought I would first implement what is needed for `metrics listen`
I will add it.
Though I think in at least the `bluechi-monitor-service` test can be done with the existing processor as well. You can listen for specific outputs during specific commands quite easily, and it would remove the need of complicated event evaluator

File: tests/bluechi_test/machine.py (hunk 1 above)
Comment (mkemel) [reply tests/bluechi_test/machine.py:132]: This code is also run during coverage collection, and then this dir is not there and is not needed. So it's not an error. But yeah, can be a warning

File: tests/bluechi_test/bluechilib/bluechictl.py
Comment (mkemel) [reply tests/bluechi_test/bluechilib/bluechictl.py:1]: As for the dir location, it is run on the controller node, but by the test runner machine. This is something internal, thus the location actually makes sense to me

Comment (mwperina) [reply tests/bluechi_test/bluechilib/bluechictl.py:1]: We just need to make sure that `bluechilib` can be imported only from python sources inside `python` subdirectory, meaning those which are executed on controller node and not those executed on test node. That's why I thought it would be great to completely separate them ...

File: tests/bluechi_test/bluechilib/bluechictl.py (hunk 2 above)
Comment (mkemel) [reply tests/bluechi_test/bluechilib/bluechictl.py]: Done

File: tests/bluechi_test/machine.py (hunk 1 above)
Comment (mwperina) [reply tests/bluechi_test/machine.py:132]: Ahh, yeah, maybe we need additional configuration and/or separate container for code coverage report generation ....

So let's just make a code comment, that this message is relevant only for `finish` phase, where code coverage report is being created

Comment (mkemel) [reply tests/bluechi_test/machine.py:132]: Done

File: tests/bluechi_test/bluechilib/bluechictl.py (hunk 2 above)
Comment (mwperina) [reply tests/bluechi_test/bluechilib/bluechictl.py]: Maybe, but having a separate class with event evaluation makes it more understandable, versatile and also more separated from the BlueChiCtlBackgroundRunner class implementation ...

File: tests/bluechi_test/bluechilib/bluechictl.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+    def seek(self):
+        self.file_desc.seek(self.pos)
+
+    def new_lines(self):
+        self.seek()
+        return '\n' in self.file_desc.read()
+
+
+class EventEvaluator:
```
Comment (mwperina) [inline tests/bluechi_test/bluechilib/bluechictl.py]: Here the class should have empty methods:

```python
class EventEvaluator:

    def process_line(self, line: str) -> None:
        pass

    def processing_finished(self) -> bool:
        pass
```

The real code should be only in relevant tests:

```python
class MetricsEventEvaluator(EventEvaluator):
    def __init__(self, expected_patterns: Set[Pattern]):
        self.expected_patterns = copy.copy(expected_patterns)
        self.all_found = False

    def process_line(self, line: str) -> None:
        for pattern in copy.copy(self.expected_patterns):
            if pattern.match(line):
                print(f"Matched line '{line}'")
                self.expected_patterns.remove(pattern)
                break

    def processing_finished(self) -> bool:
        return len(self.expected_patterns) == 0
```

File: tests/bluechi_test/bluechilib/bluechictl.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+                self.bluechictl_proc = subprocess.Popen(self.command, stdout=out_file, bufsize=1)
+
+                with FileFollower(out_file.name) as bluechictl_out:
+                    while self.bluechictl_proc.poll() is None and not self.finished:
+                        for line in bluechictl_out:
+                            print(f"Evaluating line '{line}'")
+                            self.evaluator.process_line(line)
+                            if self.evaluator.processing_finished():
+                                print("BluechiCtl output evaluation successfully finished")
+                                self.all_found = True
```
Comment (mwperina) [inline tests/bluechi_test/bluechilib/bluechictl.py]: Do we need `all_found` here? Should test evaluate `EventEvaluator` instance directly instead?

File: tests/bluechi_test/bluechilib/bluechictl.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+
+class BluechiCtlBackgroundRunner:
+    def __init__(self, command: List, evaluator: EventEvaluator):
+        self.command = command
+        self.evaluator = evaluator
+        self.thread = threading.Thread(target=self.process_events)
+        self.failsafe_thread = threading.Thread(target=self.timeout_guard, daemon=True)
+        self.all_found = False
+        self.finished = False
```
Comment (mwperina) [inline tests/bluechi_test/bluechilib/bluechictl.py]: Do we need `finished` here? Isn't it better to use `self.evaluator.processing_finished()` instead?

File: tests/tests/tier0/bluechictl-monitor-unit-created-removed/python/monitor.py
Comment (mwperina) [inline tests/tests/tier0/bluechictl-monitor-unit-created-removed/python/monitor.py:1]: Shouldn't `flake8` check formatting? If not, what do we need to configure to have the same python file formatting for the whole project?

File: tests/bluechi_test/bluechilib/bluechictl.py (hunk 3 above)
Comment (mkemel) [reply tests/bluechi_test/bluechilib/bluechictl.py]: I would keep a simple class in the library, it is reusable

File: tests/bluechi_test/bluechilib/bluechictl.py (hunk 4 above)
Comment (mkemel) [reply tests/bluechi_test/bluechilib/bluechictl.py]: right

File: tests/bluechi_test/bluechilib/bluechictl.py (hunk 5 above)
Comment (mkemel) [reply tests/bluechi_test/bluechilib/bluechictl.py]: `finished` is also used to stop the processing externally

File: tests/tests/tier0/bluechictl-monitor-unit-created-removed/python/monitor.py
Comment (mkemel) [reply tests/tests/tier0/bluechictl-monitor-unit-created-removed/python/monitor.py:1]: I used autopep and it... did stuff

File: tests/bluechi_test/bluechilib/bluechictl.py (hunk 2 above)
Comment (mkemel) [reply tests/bluechi_test/bluechilib/bluechictl.py]: You are right, the code actually looks much nicer with it.

File: tests/bluechi_test/bluechilib/bluechictl.py
Comment (mkemel) [reply tests/bluechi_test/bluechilib/bluechictl.py:1]: Actually having it there allows us to have autocomplete while developing without having it in an RPM/pip package... @engelmi

File: tests/bluechi_test/bluechilib/bluechictl.py (hunk 3 above)
Comment (mkemel) [reply tests/bluechi_test/bluechilib/bluechictl.py]: Done

File: tests/bluechi_test/bluechilib/bluechictl.py (hunk 4 above)
Comment (mkemel) [reply tests/bluechi_test/bluechilib/bluechictl.py]: Done

File: tests/tests/tier0/bluechictl-monitor-unit-created-removed/python/monitor.py
Comment (mkemel) [reply tests/tests/tier0/bluechictl-monitor-unit-created-removed/python/monitor.py:1]: Created PR to update tests/README.md
https://github.com/eclipse-bluechi/bluechi/pull/832

File: tests/bluechi_test/bluechilib/bluechictl.py (hunk 5 above)
Comment (mwperina) [reply tests/bluechi_test/bluechilib/bluechictl.py]: Ah, OK, I missed that

File: tests/bluechi_test/bluechilib/bluechictl.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+        return len(self.expected_patterns) == 0
+
+
+class BluechiCtlBackgroundRunner:
+    def __init__(self, command: List, evaluator: EventEvaluator):
+        self.command = command
+        self.evaluator = evaluator
+        self.thread = threading.Thread(target=self.process_events)
+        self.failsafe_thread = threading.Thread(target=self.timeout_guard, daemon=True)
+        self.all_found = False
```
Comment (mwperina) [inline tests/bluechi_test/bluechilib/bluechictl.py]: Still needed?

File: tests/bluechi_test/bluechilib/bluechictl.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+        pass
+
+    def processing_finished(self) -> bool:
+        pass
+
+
+class SimpleEventEvaluator(EventEvaluator):
+    def __init__(self, expected_patterns: Set[Pattern]):
+        self.expected_patterns = copy.copy(expected_patterns)
+        self.all_found = False
```
Comment (mwperina) [inline tests/bluechi_test/bluechilib/bluechictl.py]: Still needed?

File: tests/bluechi_test/bluechilib/bluechictl.py (hunk 6 above)
Comment (mkemel) [reply tests/bluechi_test/bluechilib/bluechictl.py]: Nope, removed

File: tests/bluechi_test/bluechilib/bluechictl.py (hunk 7 above)
Comment (mkemel) [reply tests/bluechi_test/bluechilib/bluechictl.py]: Nope, removed

File: tests/bluechi_test/bluechilib/bluechictl.py
Comment (mkemel) [reply tests/bluechi_test/bluechilib/bluechictl.py:1]: Scratch that, it can be placed directly under `tests` directory... But I think all files in `bluechi_test` will be able to include it anyway.

Comment (engelmi) [reply tests/bluechi_test/bluechilib/bluechictl.py:1]: @mkemel Exactly. We can move the directory one level up, `tests` is the import root for local development. But I agree, maybe `bluechilib` isn't the best name here - how about `machine_scripts/controller/xyz.py` and `machine_scripts/agent/xyz.py`? Although I don't know right now if we run scripts on the agent host. 

@mwperina The basic idea is to mimic the directory structure of the local `tests/` also in the controller host (i.e. `/tmp/bluechilib` in controller and `tests/bluechilib` on local host). This enables us to use tooling such as autocomplete when developing tests since the import path is known in both cases - which makes development a lot easier.

File: tests/bluechi_test/bluechilib/bluechictl.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+        self.finished = True
+        self.thread.join()
+
+    def found_all_patterns(self):
+        return self.evaluator.processing_finished()
+
+
+class BluechiCtl:
+    def run(self, args: List) -> Tuple[str, str]:
+        command = ["bluechictl"] + args
```
Comment (engelmi) [inline tests/bluechi_test/bluechilib/bluechictl.py]:
```suggestion
        command = ["/usr/bin/bluechictl"] + args
```
We can probably reintroduce the path here. It was an error on my side removing it.

File: tests/bluechi_test/bluechilib/bluechictl.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+                self.expected_patterns.remove(pattern)
+                break
+
+    def processing_finished(self) -> bool:
+        return len(self.expected_patterns) == 0
+
+
+class BluechiCtlBackgroundRunner:
+    def __init__(self, command: List, evaluator: EventEvaluator):
+        self.command = command
```
Comment (engelmi) [inline tests/bluechi_test/bluechilib/bluechictl.py]:
```suggestion
    def __init__(self, bluechictl_args: List, evaluator: EventEvaluator):
        self.bluechictl = BluechiCtl()
        self.bluechictl_args = bluechictl_args
```
I'd reuse the `BluechiCtl` class here and fix the usage of this background runner to it - currently it could be used to run anything.

File: tests/bluechi_test/bluechilib/bluechictl.py [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+                        time.sleep(0.5)
+            finally:
+                print("Finalizing process")
+                if self.bluechictl_proc:
+                    self.bluechictl_proc.send_signal(signal.SIGINT)
+                    self.bluechictl_proc.wait()
+                    self.bluechictl_proc = None
+
+    def timeout_guard(self):
+        time.sleep(10)
```
Comment (engelmi) [inline tests/bluechi_test/bluechilib/bluechictl.py]: Instead of a fixed value, it would be good to if `BluechiCtlBackgroundRunner`'s constructor would accept a variable for it, e.g. `timeout: int`.

File: tests/bluechi_test/bluechilib/bluechictl.py [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+            if pattern.match(line):
+                print(f"Matched line '{line}'")
+                self.expected_patterns.remove(pattern)
+                break
+
+    def processing_finished(self) -> bool:
+        return len(self.expected_patterns) == 0
+
+
+class BluechiCtlBackgroundRunner:
```
Comment (engelmi) [inline tests/bluechi_test/bluechilib/bluechictl.py]: NIT: Maybe just `BackgroundRunner`? The other classes `EventEvaluator` etc. also don't have the `BluechiCtl` prefix and they are located in the bluechictl.py.

File: tests/bluechi_test/bluechilib/bluechictl.py
Comment (mkemel) [reply tests/bluechi_test/bluechilib/bluechictl.py:1]: Don't think that there's a sense in splitting that. Also, it's more a lib than scripts, the script is run with `run_python` and uses modules from this dir

Comment (mkemel) [reply tests/bluechi_test/bluechilib/bluechictl.py:1]: `machine_lib/xyz.py` - WDYT?

Comment (mkemel) [reply tests/bluechi_test/bluechilib/bluechictl.py:1]: I still like `bluechi_machine_lib` the most

File: tests/bluechi_test/machine.py [hunk 12]
```diff
@@ -122,6 +122,22 @@ def copy_container_script(self, script_file_name: str):
...
+        curr_dir = os.getcwd()
+        source_dir = os.path.join(curr_dir, "..", "..", "..", "bluechi_test", "bluechilib")
+        if not os.path.isdir(source_dir):
+            # This message is relevant for `finish` phase, where code coverage report is being created
+            LOGGER.info("bluechilib directory not found, proceeding")
+            return
+        self.exec_run("mkdir -p /tmp/bluechi_test/bluechilib")
+        for filename in os.listdir(source_dir):
+            source_path = os.path.join(source_dir, filename)
+            if os.path.isfile(source_path):
```
Comment (engelmi) [inline tests/bluechi_test/machine.py]:
```suggestion
            if os.path.isfile(source_path) and source_path.endswith(".py"):
```
I think it doesn't hurt to copy other files as well, but we are interested only in python files.

File: tests/bluechi_test/machine.py [hunk 13]
```diff
@@ -122,6 +122,22 @@ def copy_container_script(self, script_file_name: str):
...
 
+    def copy_machine_lib(self):
+        LOGGER.info("Copying machine lib...")
+        curr_dir = os.getcwd()
+        source_dir = os.path.join(curr_dir, "..", "..", "..", "bluechi_test", "bluechilib")
+        if not os.path.isdir(source_dir):
+            # This message is relevant for `finish` phase, where code coverage report is being created
+            LOGGER.info("bluechilib directory not found, proceeding")
+            return
+        self.exec_run("mkdir -p /tmp/bluechi_test/bluechilib")
```
Comment (engelmi) [inline tests/bluechi_test/machine.py]: For a container setup it is necessary to copy the files each time. For the multihost setup we reuse hosts and, since we don't (and shouldn't) cleanup these files - we could do a check if the directory already exists and skip if it does.

File: tests/bluechi_test/bluechilib/bluechictl.py
Comment (engelmi) [reply tests/bluechi_test/bluechilib/bluechictl.py:1]: Sounds good to me :)

File: tests/bluechi_test/machine.py [hunk 14]
```diff
@@ -122,6 +122,22 @@ def copy_container_script(self, script_file_name: str):
              with content:\n{content}")
         self.create_file(target_dir, script_file_name, content)
 
+    def copy_machine_lib(self):
+        LOGGER.info("Copying machine lib...")
+        curr_dir = os.getcwd()
+        source_dir = os.path.join(curr_dir, "..", "..", "..", "bluechi_test", "bluechilib")
```
Comment (engelmi) [inline tests/bluechi_test/machine.py]: Assuming the lib would be in `tests/bluechi_machine_lib` we could also do:
```suggestion
        import bluechi_machine_lib as bml
        source_dir = bml.__path__[0]
        # ...
        # 
        target_dir = os.path.join("/", "tmp", bml.__name__)
```
Then we don't need any relative paths and we are sure that the module (`bluechi_machine_lib` in that case) exists. It is more complex, of course.

File: tests/bluechi_test/bluechilib/bluechictl.py (hunk 8 above)
Comment (mkemel) [reply tests/bluechi_test/bluechilib/bluechictl.py]: Forgot to roll back this change

File: tests/bluechi_test/bluechilib/bluechictl.py (hunk 9 above)
Comment (mkemel) [reply tests/bluechi_test/bluechilib/bluechictl.py]: The run is different in BluechiCtl and in the background runner.
But I can ensure the background runner runs only bluechictl

File: tests/bluechi_test/machine.py (hunk 14 above)
Comment (mkemel) [reply tests/bluechi_test/machine.py]: Nice

Comment (mkemel) [reply tests/bluechi_test/machine.py]: Done

File: tests/bluechi_test/machine.py (hunk 13 above)
Comment (mkemel) [reply tests/bluechi_test/machine.py]: Done

File: tests/bluechi_test/machine.py (hunk 12 above)
Comment (mkemel) [reply tests/bluechi_test/machine.py]: Done

File: tests/bluechi_test/bluechilib/bluechictl.py (hunk 11 above)
Comment (mkemel) [reply tests/bluechi_test/bluechilib/bluechictl.py]: DOne

File: tests/bluechi_test/bluechilib/bluechictl.py (hunk 10 above)
Comment (mkemel) [reply tests/bluechi_test/bluechilib/bluechictl.py]: Done

File: tests/bluechi_test/bluechilib/bluechictl.py
Comment (mkemel) [reply tests/bluechi_test/bluechilib/bluechictl.py:1]: Done

File: tests/bluechi_test/machine.py (hunk 13 above)
Comment (mkemel) [reply tests/bluechi_test/machine.py]: Done

File: tests/tests/tier0/bluechictl-metrics/python/bluechictl_metrics.py [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+    def test_bluechictl_metrics(self):
+        self.bluechictl.metrics_enable()
+        self.bgrunner.start()
+        time.sleep(0.5)
+        self.bluechictl.unit_start(node_foo_name, service_simple)
+        time.sleep(0.5)
+        self.bluechictl.unit_stop(node_foo_name, service_simple)
+        time.sleep(0.5)
+        self.bgrunner.stop()
+        assert self.bgrunner.found_all_patterns()
```
Comment (mwperina) [inline tests/tests/tier0/bluechictl-metrics/python/bluechictl_metrics.py]: Wouldn't it be cleaner to save  `SimpleEventEvaluator` instance created on line 23 and just evaluate `evaluator.processing_finished()` here? That way we could remove `found_all_patterns()` method from `BackgroundRunner` where its name looks strange

File: tests/tests/tier0/bluechictl-metrics/python/bluechictl_metrics.py [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+        args = ["metrics", "listen"]
+        self.bgrunner = BackgroundRunner(args, SimpleEventEvaluator(expected_patterns))
+        self.bluechictl = BluechiCtl()
+
+    def test_bluechictl_metrics(self):
+        self.bluechictl.metrics_enable()
+        self.bgrunner.start()
+        time.sleep(0.5)
+        self.bluechictl.unit_start(node_foo_name, service_simple)
+        time.sleep(0.5)
```
Comment (mwperina) [inline tests/tests/tier0/bluechictl-metrics/python/bluechictl_metrics.py]: Do we need the timeout between start and stop events?

File: tests/tests/tier0/bluechictl-metrics/python/bluechictl_metrics.py [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+        self.bluechictl.unit_stop(node_foo_name, service_simple)
+        time.sleep(0.5)
+        self.bgrunner.stop()
```
Comment (mwperina) [inline tests/tests/tier0/bluechictl-metrics/python/bluechictl_metrics.py:31]: Hmm, wouldn't it be better to not set `self.finished = True` in `BackgroundRunner.stop()`? It seems to me that you need to set timeout before `bgrunner.stop()` only because you stop event processing loop. But shouldn't the event processing loop be stopped only with either "all required event found" or by `timeout_guard()`?

File: tests/tests/tier0/bluechictl-monitor-unit-created-removed/python/monitor.py
Comment (mwperina) [inline tests/tests/tier0/bluechictl-monitor-unit-created-removed/python/monitor.py:1]: Could you please undo this unrelated change?

File: tests/tests/tier0/bluechi-restart-unit/python/monitor.py
Comment (mwperina) [inline tests/tests/tier0/bluechi-restart-unit/python/monitor.py:1]: Could you please undo this unrelated change?

File: tests/tests/tier0/bluechictl-metrics/python/bluechictl_metrics.py (hunk 17 above)
Comment (mkemel) [reply tests/tests/tier0/bluechictl-metrics/python/bluechictl_metrics.py:31]: I think the `timeout` should be saved to a failure of some sort, it's an overall timeout. Other than that, I want to be able to stop following output and see if I have or have not got the output I have expected or expected not to show

File: tests/tests/tier0/bluechictl-metrics/python/bluechictl_metrics.py (hunk 16 above)
Comment (mkemel) [reply tests/tests/tier0/bluechictl-metrics/python/bluechictl_metrics.py]: What do you mean? You're not talking about `time sleep`s

File: tests/tests/tier0/bluechictl-metrics/python/bluechictl_metrics.py (hunk 17 above)
Comment (mwperina) [reply tests/tests/tier0/bluechictl-metrics/python/bluechictl_metrics.py:31]: You have no idea when events will be received, that's which we have the timeout_guard, which will be raised before the global test timeout, so interrrupting processing with timeout_guard will be able to collect all the output (because global test timeout will not preserve output from a process running on a controller).

File: tests/tests/tier0/bluechictl-metrics/python/bluechictl_metrics.py (hunk 15 above)
Comment (mkemel) [reply tests/tests/tier0/bluechictl-metrics/python/bluechictl_metrics.py]: Done

File: tests/tests/tier0/bluechictl-metrics/python/bluechictl_metrics.py (hunk 16 above)
Comment (mkemel) [reply tests/tests/tier0/bluechictl-metrics/python/bluechictl_metrics.py]: Removed

File: tests/tests/tier0/bluechictl-metrics/python/bluechictl_metrics.py (hunk 17 above)
Comment (mkemel) [reply tests/tests/tier0/bluechictl-metrics/python/bluechictl_metrics.py:31]: Understood. Done

**Review by mwperina — APPROVED** [review]
LGTM

**Review by engelmi — APPROVED** [review]
LGTM
